File: Lib/Telerik/Examples/RichTextEditor/SearchHighlight/HighlightSearchedWordLayer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using Telerik.WinForms.Documents;
using Telerik.WinForms.Documents.DocumentStructure;
using Telerik.WinForms.Documents.Layout;
using Telerik.WinForms.Documents.TextSearch;
using Telerik.WinForms.Documents.UI.Layers;

namespace RichTextEditor.SearchHighlight
{
    public class HighlightSearchedWordLayer : DecorationUILayerBase
    {
        private static readonly double StrokeThikness = 1.0;
        private static readonly Brush StrokeBrush = new SolidBrush(Color.FromArgb(140, 180, 180, 0));
        private static readonly Brush FillBrush = new SolidBrush(Color.FromArgb(50, 255, 255, 0));

        public override string Name
        {
            get
            {
                return "HighlightSearchedWordLayer";
            }
        }

        private static string word;
        public static string Word
        {
            get
            {
                return word;
            }
            set
            {
                if (word != value)
                {
                    word = Regex.Escape(value);
                }
            }
        }
        public HighlightSearchedWordLayer()
        {

        }
        public override void UpdateUIViewPortOverride(UILayerUpdateContext context)
        {
      
            if (string.IsNullOrEmpty(word))
            {
                return;
            }

            this.HighlightOccurrencesInVisibleBoxes(context.GetVisibleTextLayoutBoxes());
        }

        private void HighlightOccurrencesInVisibleBoxes(IEnumerable<SpanLayoutBox> spanList)
        {
            if (spanList.Count() == 0)
            {
                return;
            }
            SpanLayoutBox firstBox = spanList.First();
            SpanLayoutBox lastBox = spanList.Last();

            DocumentPosition searchStart = new DocumentPosition(this.Document);
            DocumentPosition searchEnd = new DocumentPosition(this.Document);
            searchStart.MoveToInline(firstBox, 0);
            searchEnd.MoveToInline(lastBox, lastBox.PositionsCountInBox - 1);

            DocumentTextSearch textSearch = new DocumentTextSearch(this.Document);

            TextRange textRange = textSearch.Find(word, searchStart, searchEnd);
            int count = 0;
            while (textRange != null)
            {
                count++;
                DocumentPosition lineStart = new DocumentPosition(textRange.StartPosition);
                DocumentPosition lineEnd = new DocumentPosition(lineStart);
                lineEnd.MoveToCurrentLineEnd();
                while (lineEnd < textRange.EndPosition)
                {
                    this.FlushBoxes(lineStart, lineEnd);
                    lineStart.MoveToCurrentLineEnd();
                    lineStart.MoveToNext();
                    lineEnd.MoveToNext();
                    lineEnd.MoveToCurrentLineEnd();
                }
                this.FlushBoxes(lineStart, textRange.EndPosition);

                searchStart.MoveToPosition(textRange.EndPosition);
                if (searchStart >= searchEnd)
                {
                    break;
                }
                textRange = textSearch.Find(word, searchStart, searchEnd);                
            }
        }

        private void FlushBoxes(DocumentPosition startPosition, DocumentPosition endPosition)
        {
            if (startPosition == endPosition)
            {
                return;
            }
            RectangleF rect = new RectangleF();

            rect.X = startPosition.Location.X;
            rect.Width = endPosition.Location.X - startPosition.Location.X;

            InlineLayoutBox startBox = startPosition.GetCurrentInlineBox();
            InlineLayoutBox endBox = endPosition.GetCurrentInlineBox();

            float top = startBox.ControlBoundingRectangle.Top;
            float bottom = startBox.ControlBoundingRectangle.Bottom;

            while (startBox != endBox)
            {
                startBox = DocumentStructureCollection.GetNextSiblingForDocumentElementOnSameLevel(startBox, startBox.AssociatedDocumentElement) as InlineLayoutBox;
                if (startBox == null)
                {
                    break;
                }
                if (startBox.ControlBoundingRectangle.Top < top)
                {
                    top = startBox.ControlBoundingRectangle.Top;
                }
                if (startBox.ControlBoundingRectangle.Bottom < bottom)
                {
                    bottom = startBox.ClippedControlBoundingRectangle.Bottom;
                }
            }
            rect.Y = top;
            rect.Height = bottom - top;

            this.AddRectangle(rect);
        }

        private void AddRectangle(RectangleF rect)
        {
            var rectangle = new Telerik.WinControls.RichTextEditor.UI.Rectangle();
            rectangle.Width = Math.Max(0, rect.Width + 2 * StrokeThikness);
            rectangle.Height = rect.Height + 2 * StrokeThikness;

            rectangle.Fill = FillBrush;
            rectangle.Stroke = StrokeBrush;
            rectangle.StrokeThickness = StrokeThikness;

            Telerik.WinControls.RichTextEditor.UI.Canvas.SetTop(rectangle, rect.Top - StrokeThikness);
            Telerik.WinControls.RichTextEditor.UI.Canvas.SetLeft(rectangle, rect.Left - StrokeThikness);
            base.AddDecorationElement(rectangle);
        }
    }
}

File: Lib/Telerik/Examples/RichTextEditor/SearchHighlight/HighlightSearchedWordLayer.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Text.RegularExpressions
Imports Telerik.WinForms.Documents
Imports Telerik.WinForms.Documents.DocumentStructure
Imports Telerik.WinForms.Documents.Layout
Imports Telerik.WinForms.Documents.TextSearch
Imports Telerik.WinForms.Documents.UI.Layers

Namespace RichTextEditor.SearchHighlight
	Public Class HighlightSearchedWordLayer
		Inherits DecorationUILayerBase
		Private Shared ReadOnly StrokeThikness As Double = 1.0
		Private Shared ReadOnly StrokeBrush As Brush = New SolidBrush(Color.FromArgb(140, 180, 180, 0))
		Private Shared ReadOnly FillBrush As Brush = New SolidBrush(Color.FromArgb(50, 255, 255, 0))

		Public Overrides ReadOnly Property Name() As String
			Get
				Return "HighlightSearchedWordLayer"
			End Get
		End Property

		Private Shared word_Renamed As String
		Public Shared Property Word() As String
			Get
				Return word_Renamed
			End Get
			Set
				If word_Renamed <> Value Then
					word_Renamed = Regex.Escape(Value)
				End If
			End Set
		End Property
		Public Sub New()

		End Sub
		Public Overrides Sub UpdateUIViewPortOverride(ByVal context As UILayerUpdateContext)

			If String.IsNullOrEmpty(word_Renamed) Then
				Return
			End If

			Me.HighlightOccurrencesInVisibleBoxes(context.GetVisibleTextLayoutBoxes())
		End Sub

		Private Sub HighlightOccurrencesInVisibleBoxes(ByVal spanList As IEnumerable(Of SpanLayoutBox))
			If spanList.Count() = 0 Then
				Return
			End If
			Dim firstBox As SpanLayoutBox = spanList.First()
			Dim lastBox As SpanLayoutBox = spanList.Last()

			Dim searchStart As DocumentPosition = New DocumentPosition(Me.Document)
			Dim searchEnd As DocumentPosition = New DocumentPosition(Me.Document)
			searchStart.MoveToInline(firstBox, 0)
			searchEnd.MoveToInline(lastBox, lastBox.PositionsCountInBox - 1)

			Dim textSearch As DocumentTextSearch = New DocumentTextSearch(Me.Document)

			Dim textRange As TextRange = textSearch.Find(word_Renamed, searchStart, searchEnd)
			Dim count As Integer = 0
			Do While Not textRange Is Nothing
				count += 1
				Dim lineStart As DocumentPosition = New DocumentPosition(textRange.StartPosition)
				Dim lineEnd As DocumentPosition = New DocumentPosition(lineStart)
				lineEnd.MoveToCurrentLineEnd()
				Do While lineEnd < textRange.EndPosition
					Me.FlushBoxes(lineStart, lineEnd)
					lineStart.MoveToCurrentLineEnd()
					lineStart.MoveToNext()
					lineEnd.MoveToNext()
					lineEnd.MoveToCurrentLineEnd()
				Loop
				Me.FlushBoxes(lineStart, textRange.EndPosition)

				searchStart.MoveToPosition(textRange.EndPosition)
				If searchStart >= searchEnd Then
					Exit Do
				End If
				textRange = textSearch.Find(word_Renamed, searchStart, searchEnd)
			Loop
		End Sub

		Private Sub FlushBoxes(ByVal startPosition As DocumentPosition, ByVal endPosition As DocumentPosition)
			If startPosition Is endPosition Then
				Return
			End If
			Dim rect As RectangleF = New RectangleF()

			rect.X = startPosition.Location.X
			rect.Width = endPosition.Location.X - startPosition.Location.X

			Dim startBox As InlineLayoutBox = startPosition.GetCurrentInlineBox()
			Dim endBox As InlineLayoutBox = endPosition.GetCurrentInlineBox()

			Dim top As Single = startBox.ControlBoundingRectangle.Top
			Dim bottom As Single = startBox.ControlBoundingRectangle.Bottom

			Do While Not startBox Is endBox
				startBox = TryCast(DocumentStructureCollection.GetNextSiblingForDocumentElementOnSameLevel(startBox, startBox.AssociatedDocumentElement), InlineLayoutBox)
				If startBox Is Nothing Then
					Exit Do
				End If
				If startBox.ControlBoundingRectangle.Top < top Then
					top = startBox.ControlBoundingRectangle.Top
				End If
				If startBox.ControlBoundingRectangle.Bottom < bottom Then
					bottom = startBox.ClippedControlBoundingRectangle.Bottom
				End If
			Loop
			rect.Y = top
			rect.Height = bottom - top

			Me.AddRectangle(rect)
		End Sub

		Private Sub AddRectangle(ByVal rect As RectangleF)
			Dim rectangle = New Telerik.WinControls.RichTextEditor.UI.Rectangle()
			rectangle.Width = Math.Max(0, rect.Width + 2 * StrokeThikness)
			rectangle.Height = rect.Height + 2 * StrokeThikness

			rectangle.Fill = FillBrush
			rectangle.Stroke = StrokeBrush
			rectangle.StrokeThickness = StrokeThikness

			Telerik.WinControls.RichTextEditor.UI.Canvas.SetTop(rectangle, rect.Top - StrokeThikness)
			Telerik.WinControls.RichTextEditor.UI.Canvas.SetLeft(rectangle, rect.Left - StrokeThikness)
			MyBase.AddDecorationElement(rectangle)
		End Sub
	End Class
End Namespace
